### PR TITLE
gh-85253: Exclude `__pycache__` directories from backups using `CACHEDIR.TAG`

### PR DESCRIPTION
**Pull request by jstasiak**
This helps to prevent bloating backups with unwanted (possibly
frequently changing, if \_\_pycache__ directories are removed and
recreated) files.

Granted, one can add \_\_pycache__ directory to their
backup-tool-of-choice exclusion list, but those lists are ever
growing and maybe it'd be good to help the tools and the users.

See https://bford.info/cachedir/ for more information about the
specification – it's supported by Borg, restic, GNU Tar and attic
out of the box, other pieces of software (like rsync, Bacula and
rdiff-backup) can be made to use it (partially, just the existence of
the tag file is used, not its content)

https://bugs.python.org/issue41081

* Issue: gh-85253

### CODE REVIEW

File: Lib/importlib/_bootstrap_external.py [hunk 1]
```diff
@@ -1033,6 +1039,21 @@ def set_data(self, path, data, *, _mode=0o666):
...
 
+def _exclude_directory_from_backups(directory):
+    """Exclude the directory from backups using a CACHEDIR.TAG file.
+
+    If the file exists or there's a permission error this is a no-op.
+    """
+    try:
+        with _io.FileIO(_path_join(directory, 'CACHEDIR.TAG'), 'x') as f:
+            f.write(b"""Signature: 8a477f597d28d172789f06886806bc55
+# This file is a cache directory tag automtically created by CPython.
```
Comment (merwok) [inline Lib/importlib/_bootstrap_external.py]:
```suggestion
# This file is a cache directory tag automatically created by CPython.
```

I am not sure if this should say Python or CPython; do other implementations create pyc directories?

File: Lib/importlib/_bootstrap_external.py [hunk 2]
```diff
@@ -1033,6 +1039,21 @@ def set_data(self, path, data, *, _mode=0o666):
                                         exc)
 
 
+def _exclude_directory_from_backups(directory):
+    """Exclude the directory from backups using a CACHEDIR.TAG file.
+
+    If the file exists or there's a permission error this is a no-op.
+    """
+    try:
+        with _io.FileIO(_path_join(directory, 'CACHEDIR.TAG'), 'x') as f:
```
Comment (merwok) [inline Lib/importlib/_bootstrap_external.py:1271]: Why not use `open` here?

File: Lib/importlib/_bootstrap_external.py (hunk 1 above)
Comment (jstasiak) [reply Lib/importlib/_bootstrap_external.py]: PyPy3 creates ``__pycache__`` and ``.pyc`` files, not sure about others. I'm in no way attached to CPython in this message, could be Python just as well.

File: Lib/importlib/_bootstrap_external.py (hunk 2 above)
Comment (jstasiak) [reply Lib/importlib/_bootstrap_external.py:1271]: If I understand correctly `open` is not in scope here as it's a kind of special module that gets selected few names injected into its namespace.

**Comment by brettcannon** [comment]
FYI I don't think the issue for this has reached a conclusion that this is something we want to do yet.

File: Lib/importlib/_bootstrap_external.py [hunk 3]
```diff
@@ -1256,6 +1262,21 @@ def set_data(self, path, data, *, _mode=0o666):
...
+            f.write(b"""Signature: 8a477f597d28d172789f06886806bc55
+# This file is a cache directory tag automatically created by CPython.
```
Comment (RazerM) [inline Lib/importlib/_bootstrap_external.py]: Not that it particularly matters, but I was curious why this doesn't match the comment given on https://bford.info/cachedir/. The "automatically" seems to be from mypy. Compare the following searches:

https://github.com/search?type=code&q=%22This+file+is+a+cache+directory+tag+created+by%22
https://github.com/search?type=code&q=%22This+file+is+a+cache+directory+tag+automatically+created+by%22

```suggestion
# This file is a cache directory tag created by CPython.
```

File: Lib/importlib/_bootstrap_external.py [hunk 4]
```diff
@@ -1256,6 +1262,21 @@ def set_data(self, path, data, *, _mode=0o666):
...
+# For information about cache directory tags see https://bford.info/cachedir/
```
Comment (AA-Turner) [inline Lib/importlib/_bootstrap_external.py]: Similarly to Frazer's note above.
```suggestion
# For information about cache directory tags, see: https://bford.info/cachedir/
```

File: Lib/importlib/_bootstrap_external.py (hunk 3 above)
Comment (jstasiak) [reply Lib/importlib/_bootstrap_external.py]: I'm curious too but I can't remember now. Fair point.

**Comment by brettcannon** [comment]
Closing as I just closed the issue.

**Pull request closed without merging**